code/mission/missioncampaign.cpp

```diff
@@ -370,39 +370,42 @@ void mission_campaign_get_sw_info()
 {
     int i, count, ship_list[MAX_SHIP_CLASSES], weapon_list[MAX_WEAPON_TYPES];
 
-    // set allowable ships to the SIF_PLAYER_SHIPs
-    memset(Campaign.ships_allowed, 0, sizeof(Campaign.ships_allowed));
-    for (auto it = Ship_info.cbegin(); it != Ship_info.cend(); ++it) {
-        if (it->flags[Ship::Info_Flags::Player_ship])
-            Campaign.ships_allowed[std::distance(Ship_info.cbegin(), it)] = 1;
-    }
-
-    for (i = 0; i < MAX_WEAPON_TYPES; i++)
-        Campaign.weapons_allowed[i] = 1;
+	memset(Campaign.ships_allowed, 0, sizeof(Campaign.ships_allowed));
+	memset(Campaign.weapons_allowed, 0, sizeof(Campaign.weapons_allowed));
 
     if (optional_string("+Starting Ships:")) {
-        for (i = 0; i < static_cast<int>(Ship_info.size()); i++)
-            Campaign.ships_allowed[i] = 0;
-
         count = stuff_int_list(ship_list, MAX_SHIP_CLASSES, SHIP_INFO_TYPE);
 
         // now set the array elements stating which ships we are allowed
         for (i = 0; i < count; i++) {
             if (Ship_info[ship_list[i]].flags[Ship::Info_Flags::Player_ship])
                 Campaign.ships_allowed[ship_list[i]] = 1;
         }
-    }
+	}
+	else {
+		// set allowable ships to the SIF_PLAYER_SHIPs
+		for (auto it = Ship_info.cbegin(); it != Ship_info.cend(); ++it) {
+			if (it->flags[Ship::Info_Flags::Player_ship])
+				Campaign.ships_allowed[std::distance(Ship_info.cbegin(), it)] = 1;
+		}
+	}
 
     if (optional_string("+Starting Weapons:")) {
-        for (i = 0; i < MAX_WEAPON_TYPES; i++)
-            Campaign.weapons_allowed[i] = 0;
-
         count = stuff_int_list(weapon_list, MAX_WEAPON_TYPES, WEAPON_POOL_TYPE);
 
         // now set the array elements stating which ships we are allowed
-        for (i = 0; i < count; i++)
-            Campaign.weapons_allowed[weapon_list[i]] = 1;
+		for (i = 0; i < count; i++) {
+			if (Weapon_info[weapon_list[i]].wi_flags[Weapon::Info_Flags::Player_allowed])
+				Campaign.weapons_allowed[weapon_list[i]] = 1;
+		}
     }
+	else {
+		// set allowable weapons to the player-allowed ones
+		for (auto it = Weapon_info.cbegin(); it != Weapon_info.cend(); ++it) {
+			if (it->wi_flags[Weapon::Info_Flags::Player_allowed])
+				Campaign.weapons_allowed[std::distance(Weapon_info.cbegin(), it)] = 1;
+		}
+	}
```

Comment:
Is this a behavior change?

Reply:
Not in practical effect.  The previous version allowed all weapons in the entire table, but the player would have only been able to use player-allowed weapons.  This new version allows all player-allowed weapons.

Reply:
Since `weapons_allowed` gets written to the .CSG file, I'm pretty sure this *will* have an effect on in-development mods, although how much of one is another matter entirely. Since restricting the player's weapons to player-allowed ones happens elsewhere, though, I'm unsure of the need for this change.

Reply:
I doubt very much that any mods will be affected by this.  Even though both `+Starting Ships:` and `+Starting Weapons:` are optional when parsed, they are always written by the campaign editor save routine.  I'm certain that they are only optional because they were coded in the middle of FS1's development.  As such, this is cleanup to make the ship and weapon sections match.